### PR TITLE
feat(OAS): sanitize circular reference for Redocly

### PR DESCRIPTION
**Pull request by patrick-medusajs**
### What

Add OAS build step to patch known circular references that prevent Redocly from rendering the API documentation.

### Why

We've encountered crashing and loading issues with Redocly when the OAS contained circular references. We have been working around the limitation by omitting some known problematic $ref in our source OAS. We wish to move away from this strategy in order to always explicitly include $ref in our OAS.

### How

We are introducing a custom Redocly CLI plugin that will replace `$ref` by `type: object` base on a configurable set of instructions. These instructions can be modified in `docs-util/redocly/config.yaml`

We are adding a `redocly bundle` step in the current OAS build process in order to sanitize problematic circular references.

We updated the redocly-cli package version in order to ensure that plugins are supported.

### Test

We will use [Cart.payment](https://github.com/medusajs/medusa/blob/fd5c18515931aca7cb5d29530f5dd03ebaf7e07e/packages/medusa/src/models/cart.ts#L72-L74) to ensure that the new process is properly sanitizing.

* Run `yarn openapi:generate`
* Open `docs/api/store/components/schemas/Cart.yaml`
  * Expect the `payment` property to have been sanitized to `type: object`
* Run `yarn redocly preview-docs docs/api/store/openapi.yaml --config=docs-util/redocly/config.yaml`
* Visit http://127.0.0.1:8080/#tag/Cart/operation/GetCartsCart
  * In the response, expect cart.payment to not list the properties of the Payment schema.

### CODE REVIEW

**Review by patrick-medusajs — COMMENTED** [review]
self-review: I was not able to pipe the console output from the plugin when using exec(). I didn't want to spent too much time on that aspect since a new OAS build pipeline is being worked on.

File: docs-util/redocly/plugins/decorators/circular-patch.js
Comment (carlos-r-l-rodrigues) [inline docs-util/redocly/plugins/decorators/circular-patch.js:61]: `"#/components/schemas/".length` can be a `const`

File: docs-util/redocly/plugins/decorators/circular-patch.js
Comment (carlos-r-l-rodrigues) [inline docs-util/redocly/plugins/decorators/circular-patch.js:50]: I would suggest to replace these nested `if`s by early returns. I believe overall it will make it easier to read

```
if (ctx.type.name.toLowerCase() !== "schema" || !resolved?.location?.pointer) {
  return 
}
```

File: docs-util/redocly/plugins/decorators/circular-patch.js
Comment (carlos-r-l-rodrigues) [inline docs-util/redocly/plugins/decorators/circular-patch.js:56]:
```suggestion
              if (!patch.schemaToPatchPointers.includes(ctxSchemaPointer)) {
                continue
              }
```

**Review by carlos-r-l-rodrigues — COMMENTED** [review]
Nice job 🥇

**Review by carlos-r-l-rodrigues — APPROVED** [review]

File: docs-util/redocly/plugins/decorators/circular-patch.js
Comment (adrien2p) [reply docs-util/redocly/plugins/decorators/circular-patch.js:50]: +1

File: docs-util/redocly/plugins/decorators/circular-patch.js
Comment (adrien2p) [reply docs-util/redocly/plugins/decorators/circular-patch.js:61]: Will be easier to read as well 💪

**Review by adrien2p — APPROVED** [review]
Very nice job

File: docs-util/redocly/plugins/decorators/circular-patch.js
Comment (patrick-medusajs) [reply docs-util/redocly/plugins/decorators/circular-patch.js]: So true. Done in 34d7148997ba9a1c78dcb7b8fe3b18725c8d1f62

File: docs-util/redocly/plugins/decorators/circular-patch.js
Comment (patrick-medusajs) [reply docs-util/redocly/plugins/decorators/circular-patch.js]: Tried it. Looks better 34d7148997ba9a1c78dcb7b8fe3b18725c8d1f62

**Comment by patrick-medusajs** [comment]
Thanks for the feedback. I'll wait for @shahednasser review before merging.

**Review by shahednasser — APPROVED** [review]
LGTM! Suggestion: we should look into a way to still indicate what type of object this is. Could be as simple as including it in the description, but just something to keep in mind.

**Comment by patrick-medusajs** [comment]
> LGTM! Suggestion: we should look into a way to still indicate what type of object this is. Could be as simple as including it in the description, but just something to keep in mind.

It does if there is no description already. We could also always add it as a prefix to the description even if it already exists.
E.g.: `Payment` object. Lorem ipsum dolor sit amet.

**Comment by shahednasser** [comment]
> > LGTM! Suggestion: we should look into a way to still indicate what type of object this is. Could be as simple as including it in the description, but just something to keep in mind.
> 
> It does if there is no description already. We could also always add it as a prefix to the description even if it already exists. E.g.: `Payment` object. Lorem ipsum dolor sit amet.

Agreed on this! But I also think most of the descriptions needs revisiting from my side as they don't seem to be following a single convension, so we could leave this task until then

**Pull request merged**
